### PR TITLE
Extract Cold Paths in DocumentParser

### PR DESCRIPTION
**Pull request by original-brownbear**
Looking at the profiling of the indexing steps in the many-shards benchmarks
it seems like we don't inline much in these nested parser calls and spent quite
a bit of time on method calls.
Lets give the compiler an easier time by extracting the cold exception paths
out of this logic.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java [hunk 1]
```diff
@@ -104,6 +104,10 @@ public String documentDescription() {
         };
     }
 
+    private void throwOnLeftoverPathElements(String remainingPath) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java:107]: can it be static?

File: server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java [hunk 2]
```diff
@@ -74,9 +74,14 @@ public void addAll(List<? extends IndexableField> fields) {
     }
 
     public void add(IndexableField field) {
+        assert assertLegalFieldName(field);
+        fields.add(field);
+    }
+
+    private boolean assertLegalFieldName(IndexableField field) {
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java:81]: static?

File: server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java [hunk 3]
```diff
@@ -124,13 +124,13 @@ public Mapper build(MapperBuilderContext context) {
                 int firstDotIndex = name.indexOf(".");
                 String childName = name.substring(0, firstDotIndex);
                 String fullChildName = prefix == null ? childName : prefix + "." + childName;
-                ObjectMapper.Builder childBuilder = findChild(childName, fullChildName, context);
+                ObjectMapper.Builder childBuilder = findChild(fullChildName, context);
```
Comment (javanna) [inline server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java:127]: thanks :)

**Review by javanna — APPROVED** [review]
I left a couple of nits, LGTM otherwise. 

I am not sure I follow how especially the changes around throwing exceptions improve things, have you benchmarked before and after the change? 

Other meta question: if these do help, how do we make sure we don't lose this type of improvement? There is no guarantee that we don't go and get rid of all these additional methods in the future?

File: server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java (hunk 2 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/index/mapper/LuceneDocument.java:81]: This one uses this `this.prefix` unfortunately.

**Comment by jtibshirani** [comment]
Thanks @original-brownbear for all your work in this area. I had the same questions as @javanna. Since the code is less intuitive now, are we expecting a performance improvement? I am also worried that someone will come along and refactor this to undo the changes, since it's not obvious why it's structured this way. Is there a way we could make it more obvious what's happening, maybe by pulling some common patterns out into a utility class, and adding comments about what we're doing?

**Comment by original-brownbear** [comment]
> Is there a way we could make it more obvious what's happening, maybe by pulling some common patterns out into a utility class, and adding comments about what we're doing?

We could try to add comments on why all of these cold-paths are separate methods but it's a lot of clutter in the code.
I wonder if it would make more sense to start making more use of our JMH benchmarks. We have a few for stuff similar to this (good example is https://github.com/elastic/elasticsearch/pull/71310) but we're not actually leveraging them much. Maybe adding those kind of benchmarks and actually caring about their results is the direction we could look into for things here? (maybe a discussion for a different channel :))

**Comment by jtibshirani** [comment]
Oops, I wrote my comment at the exact same time as you. Sounds good to brainstorm together on a separate channel.

**Pull request merged**